File: LeetCode_Python/238_productOfArrayExceptSelf.py
from typing import List

class ProductOfArrayExceptSelf:
    def productOfArrayExceptSelf(self, nums : List[int]) -> List[int]:
        result = [0] * len(nums)
        
        preProduct = 1
        postProduct = 1
        
        for i in range(len(nums)):
            result[i] = preProduct
            preProduct *= nums[i]
            
        for i in range(len(nums) - 1, -1, -1):
            result[i] *= postProduct
            postProduct *= nums[i]
            
        return result
    

if __name__ == "__main__":
    solution = ProductOfArrayExceptSelf()
    
    nums = [1, 2, 3, 4]
    print(solution.productOfArrayExceptSelf(nums))
    

File: LeetCode_Java/238_productOfArrayExceptSelf.java
import java.util.Arrays;

class ProductOfArrayExceptSelf{
    public int[] productOfArrayExceptSelf(int[] nums){

        int[] result = new int[nums.length];

        int preProduct = 1;
        int postProduct = 1;

        for(int i = 0 ; i < nums.length; i++){
            result[i] = preProduct;
            preProduct = preProduct * nums[i];
        }

        for(int i = nums.length - 1; i >= 0; i--){
            result[i] = result[i] * postProduct;
            postProduct = postProduct * nums[i];
        }

        return result;
    }

    public static void main(String[] args){
        ProductOfArrayExceptSelf solution = new ProductOfArrayExceptSelf();

        int[] nums = {1, 2, 3, 4};
        int[] result = solution.productOfArrayExceptSelf(nums);

        System.out.println(Arrays.toString(result));
    }
}